handler/grpc.go

```diff
@@ -92,14 +93,70 @@ func NewGRPC(opts ...GRPCOption) (grpc.StreamHandler, io.Closer) {
 			ctx = metadata.AppendToOutgoingContext(ctx, "X-Athenz-Client-ID", c.ClientID())
 		}
 
-		conn, err := gh.dialContext(ctx, target, dialOpts...)
+		conn, err = gh.dialContext(ctx, target, dialOpts...)
 		return ctx, conn, err
 	}), gh
 }
 
+func (gh *GRPCHandler) authorize(ctx context.Context, fullMethodName string) (p authorizerd.Principal, err error) {
+	md, ok := metadata.FromIncomingContext(ctx)
+	if !ok {
+		return nil, status.Error(codes.Unauthenticated, ErrGRPCMetadataNotFound)
+	}
+	var rerr error
+	p, rerr = gh.authorizeRoleToken(ctx, fullMethodName, md)
+	if rerr == nil {
+		return p, nil
+	}
+	var aerr error
+	p, aerr = gh.authorizeAccessToken(ctx, fullMethodName, md)
+	if aerr != nil {
+		return nil, status.Error(codes.Unauthenticated, errors.Join(aerr, rerr).Error())
+	}
+	return p, nil
+}
+
+func (gh *GRPCHandler) authorizeRoleToken(ctx context.Context, fullMethodName string, md metadata.MD) (p authorizerd.Principal, err error) {
+	if !gh.roleCfg.Enable {
+		return nil, status.Error(codes.Unauthenticated, ErrRoleTokenDisabled)
+	}
+	rts := md.Get(gh.roleCfg.RoleAuthHeader)
+	if len(rts) == 0 {
+		return nil, status.Error(codes.Unauthenticated, ErrRoleTokenNotFound)
+	}
+	p, err = gh.authorizationd.AuthorizeRoleToken(ctx, rts[0], gRPC, fullMethodName)
+	if err != nil {
+		return nil, status.Error(codes.Unauthenticated, err.Error())
+	}
+	return p, nil
+}
+
+func (gh *GRPCHandler) authorizeAccessToken(ctx context.Context, fullMethodName string, md metadata.MD) (p authorizerd.Principal, err error) {
+	if !gh.atCfg.Enable {
+		return nil, status.Error(codes.Unauthenticated, ErrAccessTokenDisabled)
+	}
+	ats := md.Get(gh.atCfg.AccessTokenAuthHeader)
+	if len(ats) == 0 {
+		return nil, status.Error(codes.Unauthenticated, ErrAccessTokenNotFound)
+	}
+	cs, ok := clientCertFromContext(ctx)
+	if ok && cs != nil && cs[0] != nil {
+		p, err = gh.authorizationd.AuthorizeAccessToken(ctx, ats[0], gRPC, fullMethodName, cs[0])
+		if err != nil {
+			p, err = gh.authorizationd.AuthorizeAccessToken(ctx, ats[0], gRPC, fullMethodName, nil)
+		}
+	} else {
+		p, err = gh.authorizationd.AuthorizeAccessToken(ctx, ats[0], gRPC, fullMethodName, nil)
+	}
```

Comment:
I want this logic to be implemented similarly to the HTTP below.


https://github.com/AthenZ/athenz-authorizer/blob/65ae1a16256ae5181ba53fec8122f577b14d3065/authorizerd.go#L249-L253


Specifically, please remove the verification step after the Client Certificate-Bound Access Token verification fails, as shown below.
Performing this verification causes the system to behave as if `verifyCertThumbprint` were set to `false`, even when it is actually set to `true`.

```suggestion
	if ok && cs != nil && cs[0] != nil {
		p, err = gh.authorizationd.AuthorizeAccessToken(ctx, ats[0], gRPC, fullMethodName, cs[0])
	} else {
		p, err = gh.authorizationd.AuthorizeAccessToken(ctx, ats[0], gRPC, fullMethodName, nil)
	}
```

Reply:
fixed